Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java

```diff
@@ -0,0 +1,470 @@
+ * <a href="http://oreilly.com/catalog/9780596516246/">Algorithms in a Nutshell</a>
+ * </pre>
+ */
+public final class KDTree implements NeighboursQuery {
+
+    private final SGDVector[] data;
+    private final int numThreads;
+
+    private final DimensionNode root;
+
+    /**
+     * Constructs a k-d tree nearest neighbour query object using the supplied parameters.
+     *
+     * @param data The data that will be used for neighbour queries.
+     * @param distanceType The distance function.
+     * @param numThreads The number of threads to be used to parallelize queries for multiple points.
+     */
+    KDTree(SGDVector[] data, DistanceType distanceType, int numThreads) {
+
+
+        this.data = data;
+        this.numThreads = numThreads;
+
+        int numDimensions = data[0].size();
+
+        // The original data is copied into this array of intAndVectors. The new array will be rearranged, throughout the
+        // process of tree generation, and the original indices need to maintained. The size of each vector is confirmed
+        // to be equal.
+        IntAndVector[] points = new IntAndVector[data.length];
+        for (int i=0; i < data.length; i++) {
+            points[i] = new IntAndVector(i, data[i]);
+            if (data[i].size() != numDimensions) {
+                throw new IllegalArgumentException("All the SGDVectors must be the same size.");
+            }
+        }
+
+        // Start generating the tree at dimension 1, which is index 0 of the array of IntAndVectors.
+        root = generateTree(1, numDimensions, points, 0, data.length - 1, distanceType);
+    }
+
+    @Override
+    public List<Pair<Integer, Double>> query(SGDVector point, int k) {
+        DistanceIntAndVectorBoundedMinHeap queue = new DistanceIntAndVectorBoundedMinHeap(k);
+
+        // make a fast initialization of the queue starting with points likely close to the target.
+        initializeQueue(point, queue);
+
+        // perform a thorough traversal of the tree to find the k nearest neighbours.
+        root.nearest(point, queue, false);
+
+        @SuppressWarnings("unchecked")
+        Pair<Integer, Double>[] indexDistanceArr = (Pair<Integer, Double>[]) new Pair[k];
+        int i = 1;
+        // Use an array to put the polled items from the queue into a sorted ascending order, by distance.
+        while (!queue.isEmpty()) {
+            MutableDistIntAndVectorTuple tuple = queue.poll();
+            indexDistanceArr[k - i] = new Pair<>(tuple.intAndVector.idx, tuple.dist);
+            i++;
+        }
+        return new ArrayList<>(Arrays.asList(indexDistanceArr));
+    }
+
+    @Override
+    public List<List<Pair<Integer, Double>>> query(SGDVector[] points, int k) {
+        int numQueries = points.length;
+
+        @SuppressWarnings("unchecked")
+        List<Pair<Integer, Double>>[] indexDistancePairListArray = (List<Pair<Integer, Double>>[]) new ArrayList[numQueries];
```

Comment:
This could be `new List[numQueries]`, which might fix the issue above.

Reply:
Yep, this was the issue. Thanks for persisting on this, to achieve the cleanest code here. I've made a similar change to `NeighboursBruteForce`